koku/masu/processor/aws/aws_report_db_cleaner.py

```diff
@@ -42,6 +42,58 @@ def __init__(self, schema):
         """
         self._schema = schema
 
+    def purge_expired_line_item(self, expired_date=None, provider_uuid=None, simulate=False):
+        """Remove raw line item report data with a billing start period before specified date.
+
+        Args:
+            expired_date (datetime.datetime): The cutoff date for removing data.
+            provider_uuid (uuid): The DB id of the provider to purge data for.
+            simulate (bool): Whether to simluate the removal.
+
+        Returns:
+            ([{}]) List of dictionaries containing 'account_payer_id' and 'billing_period_start'
+
+        """
+        LOG.info("Calling purge_expired_line_item for aws")
+
+        with ReportingCommonDBAccessor() as reporting_common:
+            column_map = reporting_common.column_map
+
+        with AWSReportDBAccessor(self._schema, column_map) as accessor:
+            if (expired_date is None and provider_uuid is None) or (  # noqa: W504
+                expired_date is not None and provider_uuid is not None
+            ):
+                err = "This method must be called with either expired_date or provider_uuid"
+                raise AWSReportDBCleanerError(err)
+            removed_items = []
+
+            if expired_date is not None:
+                bill_objects = accessor.get_bill_query_before_date(expired_date)
+            else:
+                bill_objects = accessor.get_cost_entry_bills_query_by_provider(provider_uuid)
+            with schema_context(self._schema):
+                for bill in bill_objects.all():
+                    bill_id = bill.id
+                    removed_payer_account_id = bill.payer_account_id
+                    removed_billing_period_start = bill.billing_period_start
+
+                    if not simulate:
+                        del_count = accessor.get_lineitem_query_for_billid(bill_id).delete()
+                        LOG.info("Removing %s cost entry line items for bill id %s", del_count, bill_id)
```

Comment:
This approach might have performance consequences that we might want to think about. A moderately large cost report can be tens of millions of lineitem rows, which may require that we lock the table to do the deletes. Or we may need to think about table partitioning or other strategies. (Some of this might be moot if we move away from Postgres in the medium-term.)

Have you tried running this delete while masu is simultaneously trying to ingest new data?  I'm curious if we'd see a deadlock or just poor performance.

I'll defer to @adberglund or @Red-HAP , as I think they've got more DB experience than I do. But, I'd like to at least sanity-check that this is a workable solution for the size of the data we're expecting to see. :slightly_smiling_face:

Reply:
I have tried running this delete while data is being ingested and what happens is that the delete task is queued until after the ingestion is completed. As for the performance concerns, I pretty much stole this logic from here: https://github.com/project-koku/koku/blob/master/koku/masu/processor/aws/aws_report_db_cleaner.py#L86

Which means that if we do have performance concerns about this method we should make sure to do it in both places, and it seems like there a lot of delete tasks that do something very similar in the `purge_expired_report_data` method, so it may be worth turning this into its own issue at that point to check on it.

Reply:
I think the task stayed queued because celery was only set to run 1 task at a time, if we were running multiple workers it could trigger. However, the lock is a row exclusive lock, so it will only lock the rows being deleted, not prevent new data from being inserted. There is a case that we do want to prevent: This task gets called and run in the middle of processing a report from last month (like say at the beginning of this month when we get updates to the previous month's report). So we probably want some condition to make sure we only run it after finishing processing a report from the previous month.

This brings up another important point that I missed last week. We need a place that this is regularly being called. Unless I missed that, I think the only way to trigger this today is via the API call. We could add the call to the `remove_expired_data` task with `line_items_only=True` as the last item of the chain here. 

https://github.com/project-koku/koku/blob/7acb90eb68c69268ea8ca9c616e5ff2bc9efe04f/koku/masu/processor/tasks.py#L233-L238

We could conditionally trigger that task when we've just processed data from a previous month. The month of the `start_date` in the `update_summary_tables` task would not be the current month.

Reply:
Andrew and I talked about this, and we decided it would be easier to review the trigger in tasks.py as a separate pr to this user story and to use this pr as a base.

Reply:
The DELETE lock type is `ROW EXCLUSIVE` which does not block access to the table, just the deleted rows. But we should make sure we only run this during certain times